smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -84,51 +145,182 @@ class DragonBackend:
-        # dictionary maps hostname to step_id of
-        # step being executed on it
+        """ProcessGroup execution state information"""
+        self._queue_lock = RLock()
+        """Lock that needs to be acquired to access internal queues"""
+        self._step_id: int = 0
+        """Incremental ID to assign to new steps prior to execution"""
+
         self._initialize_hosts()
         self._queued_steps: "collections.OrderedDict[str, DragonRunRequest]" = (
             collections.OrderedDict()
         )
+        """Steps waiting for execution"""
+        self._stop_requests: t.Deque[DragonStopRequest] = collections.deque()
+        """Stop requests which have not been processed yet"""
         self._running_steps: t.List[str] = []
+        """List of currently running steps"""
         self._completed_steps: t.List[str] = []
-
+        """List of completed steps"""
+        self._last_beat: float = 0.0
+        """Time at which the last heartbeat was set"""
+        self._heartbeat()
+        self._last_update_time = self._last_beat
+        """Time at which the status update was printed the last time"""
         num_hosts = len(self._hosts)
-        host_string = str(num_hosts) + (" hosts" if num_hosts > 1 else " host")
+        host_string = str(num_hosts) + (" hosts" if num_hosts != 1 else " host")
         self._shutdown_requested = False
-        self._updates = 0
-        print(f"{host_string} available for execution: {self._hosts}")
-
-    def print_status(self) -> None:
-        print("\n-----------------------Launcher Status-----------------------")
-        print(f"| {self._updates}: System hosts: ", self._hosts)
-        print(f"| {self._updates}: Free hosts: ", list(self._free_hosts))
-        print(f"| {self._updates}: Allocated hosts: ", self._allocated_hosts)
-        print(f"| {self._updates}: Running steps: ", self._running_steps)
-        print(f"| {self._updates}: Group infos: ", self._group_infos)
-        print(f"| {self._updates}: There are {len(self._queued_steps)} queued steps")
-        print("-------------------------------------------------------------\n")
+        """Whether the shutdown was requested to this server"""
+        self._can_shutdown = False
+        """Whether the server can shut down"""
+        self._frontend_shutdown: bool = False
+        """Whether the server frontend should shut down when the backend does"""
+        self._shutdown_initiation_time: t.Optional[float] = None
+        """The time at which the server initiated shutdown"""
+        smartsim_config = get_config()
+        self._cooldown_period = (
+            smartsim_config.telemetry_frequency * 2 + 5
+            if smartsim_config.telemetry_enabled
+            else 5
+        )
+        """Time in seconds needed to server to complete shutdown"""
+        logger.debug(f"{host_string} available for execution: {self._hosts}")
+
+    @staticmethod
+    def _proc_group_info_table_line(
```

Comment:
I don't think formatting a table to print belongs in this class. I reiterate the idea of a View class to render details to logs.

Reply:
Created `DragonBackendView` class that does all string generation for outputs.